### PR TITLE
Enforce Standard Logging Compliance by Abstraction

### PR DESCRIPTION
**Pull request by coleFD**
The goal of this commit is to force compliance of standard logging practices so more efficient indexing features can be developed in the future such as a logs bloom filter by event name. See the original discussion here: https://github.com/near/NEPs/issues/297

! NOTE:
Please see discussion in the linked issue above to determine if it would be better to make the `standard` and `version` field optional

! BREAKING CHANGE: 
For this change to be useful, we will need to privatize/remove the `env::log()` and `env::log_str()` functions, which has not been implemented yet in this commit. May not be a breaking change, but adding this to this section, so it is recognized

### CODE REVIEW

File: near-sdk/src/environment/env.rs
Comment (frol) [inline near-sdk/src/environment/env.rs:1070]: APIs like this are error-prone, and I would not teach our users bad practices.

Can you take inspiration from [FT](https://github.com/near/near-sdk-rs/blob/770cbce018a1b6c49d58276a075ace3da96d6dc1/near-contract-standards/src/fungible_token/events.rs) / [NFT](https://github.com/near/near-sdk-rs/blob/master/near-contract-standards/src/non_fungible_token/events.rs) events and design the API that would help people to design their events reliably? E.g. encourage people to define their events as enum, and bake it into the flow somehow; here is just an illustration of how that could look like:

```rust
enum Mintbase001Event {
    Mint { owner_id: AccountId, token_id: String },
    Burn { ... },
    Transfer { ... },
}

impl JsonEventEmit for Mintbase001Event { /* there will be a default `.emit()` implementation which just serializes `&self` and logs it */ }

impl Tokens {
    fn mint(&mut self, token: Token) {
         self.insert(token);
         Mintbase001Event::Mint { ... }.emit();
    }
}

self.tokens.mint(token);
```

Comment (coleFD) [reply near-sdk/src/environment/env.rs:1070]: Sorry, I am new to open source dev. Is it best practice to `--amend` the commit since this was more of a proposal commit or to make a new commit?

Also, is a logging structure like this `Event::emit(MyEvent::Mint { ... })` acceptable, or is the structure above the only acceptable structure?

Comment (frol) [reply near-sdk/src/environment/env.rs:1070]: @coleFD we will squash PR commits here, so feel free to just push new commits

**Comment by coleFD** [comment]
Changed to WIP because I I would like to convert to a single derive attribute with kwargs, and I need to resolve issues with lifetimes. Any feedback on this pattern (see `./src/types/events` test for pattern)

File: near-sdk/src/environment/env.rs
Comment (frol) [reply near-sdk/src/environment/env.rs]: > Also, is a logging structure like this Event::emit(MyEvent::Mint { ... }) acceptable, or is the structure above the only acceptable structure?

That makes sense to me. My proposed structure was just an illustration of how to promote the events to be reliable instead of ad-hoc.

**Comment by frol** [comment]
@coleFD Exposing events via ABI would be awesome! I feel it is a much heavier lift than the initial idea, but I see it as a cool challenge :-D

**Comment by coleFD** [comment]
> near_bindgen(event)

For sure, I dont think adding events to the abi should be added to this issue. I dont want to add to many changes into 1 issue, but If I utilize `near_bindgen` instead of creating a new macro, it will put us in a better spot to add the abi functionality in the future.

File: near-sdk-macros/src/core_impl/event/mod.rs
Comment (frol) [inline near-sdk-macros/src/core_impl/event/mod.rs:27]: I feel I might be missing something, why do we need two ways of achieving the same thing? I mean there is `#[near_bindgen(events)]` and this `#[near_events]`.

File: near-sdk-macros/src/core_impl/event/mod.rs
Comment (frol) [inline near-sdk-macros/src/core_impl/event/mod.rs:45]: How does this `Event::emit {` work? Shouldn't it be a function call with `()` instead of `{}`?

**Comment by frol** [comment]
> I still need to remove the previous logging functions.

Please, clarify which logging functions you refer to.

**Comment by frol** [comment]
@austinabell Could you recommend someone to review this PR?

File: near-sdk-macros/src/core_impl/event/mod.rs
Comment (coleFD) [reply near-sdk-macros/src/core_impl/event/mod.rs:27]: I will change this. I made the conversion from `#[near_events]` to `#[near_bindgen(events)]` yesterday.

File: near-sdk-macros/src/core_impl/event/mod.rs
Comment (coleFD) [reply near-sdk-macros/src/core_impl/event/mod.rs:45]: yep, I will change this

**Comment by coleFD** [comment]
> > I still need to remove the previous logging functions.
> 
> Please, clarify which logging functions you refer to.

The goal of this is to prevent logging without having the standard metadata fields `standard`, `version`, `event`, and `data`, so any logging functionality that does not do that, I think should be removed from the users' toolbox. I think the only functions that do this are `env::log()`, `env::log_str()`, and the macro `log!()`. 

Are there any concerns with doing this?

**Comment by frol** [comment]
log_str and its derivatives should be left as is. There is no problem with using logs in contracts, though, we should definitely promote usage of events. Also, I would imagine we reimplement NFT and FT events using this facility. I see that @austinabell is out of office until next week, so we will need to wait until he gets back and gets time to review it.

**Comment by frol** [comment]
By the way, I have just learned about this prior art: https://github.com/NEARFoundation/near-contract-tools#events

**Comment by coleFD** [comment]
> By the way, I have just learned about this prior art: https://github.com/NEARFoundation/near-contract-tools#events

interesting, is there still incentive to add this to `near-sdk`?

**Comment by frol** [comment]
I would like to hear what @austinabell thinks

File: near-sdk/src/types/event.rs
Comment (austinabell) [inline near-sdk/src/types/event.rs:28]: Curious why the decision to group all events in one enum? Would this not just be simpler if it was an attribute on a struct? I suppose this is a bit opinionated, but wondering if there is a reason this _must_ be an enum

File: near-sdk-macros/src/lib.rs
Comment (austinabell) [inline near-sdk-macros/src/lib.rs:441]: Clone shouldn't be a required bound. You don't need an owned copy of the data to serialize. I haven't checked if this compiles, but it would look something like this:
```suggestion
            struct EventBuilder<'a> #impl_generics #where_clause {
                standard: String,
                version: String,
                #[serde(flatten)]
                event_data: &'a #name #type_generics
            }
            impl #impl_generics near_sdk::StandardEvent for #name #type_generics #where_clause {
                fn format(&self) -> String {
                    let (standard, version): (String, String) = match self {
                        #(#event_meta),*
                    };
                    let event = EventBuilder {standard, version, event_data: self };
                    near_sdk::serde_json::to_string(&event).unwrap_or_else(|_| near_sdk::env::abort())
                }

                fn emit(&self) {
                    near_sdk::Event::emit(self);
```

File: near-sdk/src/types/event.rs
Comment (austinabell) [inline near-sdk/src/types/event.rs:3]: This could have a default implementation to prevent devs from having to implement it manually if they choose.
```suggestion
    fn emit(&self) {
        crate::env::log_str(&format!("EVENT_JSON:{}", &standard_event.format()));
    }
```

File: near-sdk/src/types/event.rs
Comment (austinabell) [inline near-sdk/src/types/event.rs:11]: This should just be removed altogether. It's basically a duplicate of the other trait and would be confusing to have this type. 

Also, side note is you don't need an owned value of `standard_event` here

**Review by austinabell — COMMENTED** [review]
Thanks for working on this! I think this can be

File: near-sdk/src/types/event.rs
Comment (coleFD) [reply near-sdk/src/types/event.rs:11]: This was added to provide the pattern `Event::emit(event)`. Personally, I like this more than using the `.emit()` method since it is easier to locate when skimming a contract, but we can go the `.emit()` method route if yall like that more. I do agree there should only be one option to emit events.

File: near-sdk/src/types/event.rs
Comment (coleFD) [reply near-sdk/src/types/event.rs:2]: if we decide not to implement the `Event` type, I could just remove this and add it to the `.emit()` method

File: near-sdk/src/types/event.rs
Comment (austinabell) [reply near-sdk/src/types/event.rs:11]: You could still do that pattern using the other trait, though. What's the need to have it as a separate trait?

File: near-sdk/src/types/event.rs
Comment (austinabell) [reply near-sdk/src/types/event.rs:28]: nope, that's fine. It's opinionated, was just sharing my intuition. I suppose we could also just support _both_ if it doesn't increase complexity meaningfully.

cc @itegulov wdyt?

File: near-sdk/src/types/event.rs
Comment (coleFD) [reply near-sdk/src/types/event.rs:11]: okay, I see what you are saying

File: near-sdk/src/types/event.rs
Comment (frol) [reply near-sdk/src/types/event.rs:28]: I treat events as some sort of standard and an enum in my head represents the fact that there is a finite number of types of events that a contract can emit. That said, I would consider extracting `standard = "..."` to the enum declaration level instead of the enum variants level. This way we could have FtEvents, NftEvents, MtEvents, FayyrEvents, etc, so one could import and use them in a composable way.

File: near-sdk/src/types/event.rs
Comment (austinabell) [reply near-sdk/src/types/event.rs:2]: > thoughts on this interface?
> 
> ```
> pub trait StandardEvent {
>     fn format(&self) -> std::borrow::Cow<str>;
>     fn emit(&self) {
>         crate::env::log_str(&format!("EVENT_JSON:{}", self.format()));
>     }
>     fn emit_event(standard_event: &Self) {
>         crate::env::log_str(&format!("EVENT_JSON:{}", standard_event.format()));
>     }
> }
> ```

`emit_event` is the same interface as `emit`, just slightly more limiting. I don't see the point of it

Comment (coleFD) [reply near-sdk/src/types/event.rs:2]: What about returning `Cow<str>` for `format()`?

Comment (austinabell) [reply near-sdk/src/types/event.rs:2]: Oh, didn't notice. Why not just have it as an associated type bound on `std::fmt::Display` or `AsRef<str>`? Adding `Cow` code to everyone's contract, even if they don't want to use it, seems unnecessary. We currently avoid `Cow` in the SDK for that reason.

File: near-sdk-macros/src/core_impl/utils/mod.rs
Comment (itegulov) [inline near-sdk-macros/src/core_impl/utils/mod.rs:95]: nit: any reason to pre-wrap `v` as a pair? I personally find that this looks more readable:
```suggestion
pub(crate) fn get_event_args(var: &syn::Variant) -> (Option<LitStr>, Option<LitStr>) {
    let mut standard = None;
    let mut version = None;
    var.attrs.iter().for_each(|attr| {
        let test = attr.parse_args_with(
            syn::punctuated::Punctuated::<syn::MetaNameValue, syn::Token![,]>::parse_terminated,
        );
        if let Ok(item) = test {
            item.iter().for_each(|kwargs| {
                if let syn::MetaNameValue { path, lit: syn::Lit::Str(value), .. } = kwargs {
                    if path.is_ident("standard") {
                        standard = Some(value.clone());
                    } else if path.is_ident("version") {
                        version = Some(value.clone());
                    }
                };
            })
        }
    });
    (standard, version)
```

File: near-sdk-macros/src/lib.rs
Comment (itegulov) [inline near-sdk-macros/src/lib.rs:375]:
```suggestion
/// The attribute macro `near_bindgen(events)` injects this macro and should be the used over this
```

File: near-sdk-macros/src/core_impl/utils/mod.rs
Comment (itegulov) [inline near-sdk-macros/src/core_impl/utils/mod.rs:89]: Should we prohibit multiple conflicting `#[standard]`/`#[version]` attributes?

File: near-sdk-macros/src/lib.rs
Comment (itegulov) [inline near-sdk-macros/src/lib.rs:82]:
```suggestion
/// #[near_bindgen(events)]
/// pub enum MyEvents {
///    #[event_meta(standard = "swap_standard", version = "1.0.0")]
///    Swap { token_in: AccountId, token_out: AccountId, amount_in: u128, amount_out: u128 },
///
///    #[event_meta(standard = "string_standard", version = "2.0.0")]
///    StringEvent(String),
///
///    #[event_meta(standard = "empty_standard", version = "3.0.0")]
///    EmptyEvent
/// }
```

File: near-sdk-macros/src/core_impl/event/mod.rs
Comment (itegulov) [inline near-sdk-macros/src/core_impl/event/mod.rs:40]:
```suggestion
/// pub enum MyEvents {
///    #[event_meta(standard = "swap_standard", version = "1.0.0")]
///    Swap { token_in: AccountId, token_out: AccountId, amount_in: u128, amount_out: u128 },
///
///    #[event_meta(standard = "string_standard", version = "2.0.0")]
///    StringEvent(String),
///
///    #[event_meta(standard = "empty_standard", version = "3.0.0")]
///    EmptyEvent
/// }
```

File: near-sdk-macros/src/lib.rs
Comment (itegulov) [inline near-sdk-macros/src/lib.rs:447]:
```suggestion
            syn::Error::new(Span::call_site(), "EventMetadata can only be used as a derive on enums.")
```

File: near-sdk-macros/src/core_impl/utils/mod.rs
Comment (itegulov) [inline near-sdk-macros/src/core_impl/utils/mod.rs:83]: Shouldn't we be checking that `attr` is `event_meta` just in case something else has `standard`/`version` args?

File: near-sdk/src/types/event.rs
Comment (itegulov) [reply near-sdk/src/types/event.rs:28]: > @itegulov wdyt about this? Do you think the implications of enum vs struct annotation would have any implication on integrating into the ABI?

I don't think it matters, we should be able to generate ABI for both options.

Personally, I find that having a mega-enum covering all standards used in this contract is a bit of an antipattern, but not a strong opinion. What @frol suggested makes more sense to me - having an enum per standard looks cleaner.

File: near-sdk-macros/src/lib.rs
Comment (frol) [inline near-sdk-macros/src/lib.rs:363]: It was hard to understand what you mean here. I hope I managed to unpack it correctly:

```suggestion
/// NOTE: This is an internal implementation for `#[near_bindgen(events(standard = ...))]` attribute.
```

**Review by frol — APPROVED** [review]
Looks good to me.

@austinabell @itegulov What do you think?

**Comment by frol** [comment]
@coleFD CI is not happy about fmt.

@austinabell Buildkite CI failed with the following error:

image

I don't see this PR changing Cargo.* or rust toolchain in any way.

**Comment by itegulov** [comment]
> Buildkite CI failed with the following error:

This was fixed in https://github.com/near/near-sdk-rs/pull/942. Just needs to be rebased I believe.

File: near-sdk/src/types/event.rs
Comment (austinabell) [inline near-sdk/src/types/event.rs:6]: Sorry, I initially mentioned `std::fmt::Display` as an alternative to `AsRef<str>` but thinking about it more, it doesn't make much sense. This is because it forces `fmt` machinery in, even if someone wanted to have a static event be emitted.

```suggestion
    type EventString: std::convert::AsRef<str>;
    fn format(&self) -> Self::EventString;
    fn emit(&self) {
        format!("EVENT_JSON:{}", self.format().as_ref());
    }
```

File: near-sdk/src/types/event.rs
Comment (austinabell) [inline near-sdk/src/types/event.rs:2]: Wondering if anyone else has opinions, but having this named as `EventJson` might be more clear?

Honestly not thrilled that our events and API is being forced as JSON, and we don't have much flexibility to use a more efficient format in the future, but this would make it more clear what this format is?

wdyt @frol ?

Comment (frol) [reply near-sdk/src/types/event.rs:2]: Yeah, naming is hard. The standard uses `EVENT_JSON:` prefix specifically to keep the opportunity open for future extension. Thus, `EventJson` makes total sense to me as it matches the prefix. Maybe even consider renaming `#[near_bindgen(event)]` to `#[near_bindgen(event_json)]`?

Comment (frol) [reply near-sdk/src/types/event.rs:2]: I don't want to drag ourselves into the conversation of how other formats could look like - borsh does not have a schema, and binary data cannot be dumped into UTF-8 string, so before getting dumped into logs, we either would need to make a protocol change in NEAR Protocol that would turn the logs to be an array of bytes instead of a valid UTF-8, or encode borsh-serialized event into, just for the sake of example, base64.

Thus, I gravitate towards actually renaming all the mentions of just `event` into `event_json` and get back to this discussion later, when we will have several events standards.

Comment (coleFD) [reply near-sdk/src/types/event.rs:2]: should I move forward with @frol recommendation?

Comment (austinabell) [reply near-sdk/src/types/event.rs:2]: I don't have an opinion if you're asking me. What frol suggested SGTM :)

**Comment by mooori** [comment]
Using [`darling`](https://github.com/TedDriggs/darling) to parse attributes passed to `#[near_bindgen]` could help to remove custom parsing logic.

Another benefit of using `darling` is related to supporting other attributes besides `events`. That could be done by adding fields to the structs into which attributes are parsed, instead of adding more custom parsing logic.

For example, [this PR](https://github.com/near/near-sdk-rs/pull/952) shows how `darling` could be used for `near_bindgen` and handling multiple attributes might [look like this](https://github.com/near/near-sdk-rs/pull/952#issuecomment-1296883879).

File: near-sdk-macros/src/lib.rs
Comment (austinabell) [inline near-sdk-macros/src/lib.rs:86]: since the param is used now, can remove the `_` prefix

File: near-sdk/src/types/event.rs
Comment (austinabell) [inline near-sdk/src/types/event.rs:65]: nit: this should be `assert_eq` to have a better error message if this is ever broken

File: near-sdk/src/types/event.rs
Comment (austinabell) [inline near-sdk/src/types/event.rs:2]: My previous comment was referring to changing the name of the associated type, not the trait name, but I suppose this is fine. Only thing I'll mention is that it is ambiguous what the `EventString` is referring to -- whether it be the whole event message or just the JSON data

File: near-sdk-macros/src/lib.rs
Comment (austinabell) [inline near-sdk-macros/src/lib.rs:415]: confused, if you're just erroring this static message if at least one error, why not just do it where you increment `attr_error`?

File: near-sdk-macros/src/lib.rs
Comment (austinabell) [inline near-sdk-macros/src/lib.rs:428]: This will conflict if you have two event derives in the same module. Since this is a hidden type, ideally it's scoped only for the implementation of `EventJson`. Am I missing something? Why else would it be used otherwise?

File: near-sdk-macros/src/lib.rs
Comment (austinabell) [inline near-sdk-macros/src/lib.rs:405]: why insert the lifetime param at the end? Lifetimes must come before type generics. (honestly confused as why this isn't broken in test, so might be missing something here)

File: near-sdk-macros/src/lib.rs
Comment (coleFD) [reply near-sdk-macros/src/lib.rs:428]: Yes, it's only used for the implementation of `EventJson` to wrap the user defined events enum and add the `standard` and `version` field with correct formatting.

File: near-sdk-macros/src/lib.rs
Comment (coleFD) [reply near-sdk-macros/src/lib.rs:415]: compiler errors in the `.for_each()`. I can convert to a for loop and do what you proposed

File: near-sdk-macros/src/lib.rs
Comment (coleFD) [reply near-sdk-macros/src/lib.rs:405]: `syn` handles ordering in the `ToTokens` implementation for generics. I will change this to insert it at the 0 index so it's not confusing

**Comment by coleFD** [comment]
any feedback on the latest commit?

**Comment by austinabell** [comment]
> any feedback on the latest commit?

There's actually a much simpler way to make sure there is no overlap here. I made the commit https://github.com/near/near-sdk-rs/commit/1e6c09edf1778fb6e2ee1604965df09bcc11545c.

Feel free to cherry-pick, make a similar change yourself, or let me know if I can push it to this branch. It's a small nit but with this, you can still have a collision when you don't need that to be the case.

**Comment by austinabell** [comment]
So when making that other change, it wasn't clear to me why the trait is necessary at all. Seems unergonomic to require devs to import that trait to use this. Why not just remove it like in https://github.com/near/near-sdk-rs/commit/1cc5c19d134c5e72ef3ec5cfe7e1f5701b82ed7f ?

**Comment by coleFD** [comment]
> So when making that other change, it wasn't clear to me why the trait is necessary at all. Seems unergonomic to require devs to import that trait to use this. Why not just remove it like in [1cc5c19](https://github.com/near/near-sdk-rs/commit/1cc5c19d134c5e72ef3ec5cfe7e1f5701b82ed7f) ?

For my first iteration, the plan was to have a `StandardEvent` trait to standardize the interface for all current and future serialization formats. I agree it doesnt make much sense if we just have json

**Comment by austinabell** [comment]
> For my first iteration, the plan was to have a `StandardEvent` trait to standardize the interface for all current and future serialization formats. I agree it doesn't make much sense if we just have JSON

I just don't see the benefit in having a trait for this if all it's going to do is add the convenience `env::log_str("EVENT_JSON:...")`. Can always add the trait back later, but I'd rather keep the API as slim as possible.

I'll push that change to this branch then since it seems like you're on board. Would be nice to have a macro test for the codegen, since I was able to make both changes without any tests failing, which is not ideal. This is not required though, improved tests can come after this PR comes in. Sorry for so much back and forth, this is just an important interface and it would be painful to migrate later.

**Comment by frol** [comment]
@austinabell @coleFD Thanks for pushing it forward! I was out of the office, and I just had a chance to sync back. I see there was a huge progress! 🙏

**Review by austinabell — APPROVED** [review]
LGTM but would be good to have another approval since I was the last to make a commit on this branch

File: near-sdk/src/types/mod.rs
Comment (frol) [inline near-sdk/src/types/mod.rs:17]: Is it actually the right place for the tests? Should we move it to `near-sdk/tests`?

File: near-sdk-macros/src/lib.rs
Comment (frol) [inline near-sdk-macros/src/lib.rs:80]:
```suggestion
///         MyEvents::StringEvent(
///             String::from("some_string")
///         ).emit();
```

File: near-sdk-macros/src/core_impl/event/mod.rs
Comment (frol) [inline near-sdk-macros/src/core_impl/event/mod.rs:16]:
```suggestion
                "Near events must have a `standard` value as an argument for `event_json` in the `near_bindgen` arguments. The value must be a string literal, e.g. \"nep999\", \"mintbase-marketplace\".",
```

**Review by frol — APPROVED** [review]
Overall, looks good to me. I only left a few nits in the comments

**Comment by austinabell** [comment]
thank you! Sorry for all the back and forth

**Comment by coleFD** [comment]
No worries! thank you guys for being responsive and taking the time to dig into it!

**Pull request merged**
